mlflow/projects/__init__.py

```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
+def _get_entry_point_command(project_dir, entry_point, use_conda, parameters, storage_dir):
+    project = _load_project(project_dir=project_dir)
     storage_dir_for_run = _get_storage_dir(storage_dir)
     eprint("=== Created directory %s for downloading remote URIs passed to arguments of "
            "type 'path' ===" % storage_dir_for_run)
-    # Try to build the command first in case the user mis-specified parameters
-    run_project_command = project.get_entry_point(entry_point)\
-        .compute_command(parameters, storage_dir_for_run)
     commands = []
     if use_conda:
-        conda_env_path = os.path.abspath(os.path.join(work_dir, project.conda_env))
-        _maybe_create_conda_env(conda_env_path)
+        conda_env_path = os.path.abspath(os.path.join(project_dir, project.conda_env))
         commands.append("source activate %s" % _get_conda_env_name(conda_env_path))
+    commands.append(
+        project.get_entry_point(entry_point).compute_command(parameters, storage_dir_for_run))
+    return " && ".join(commands)
 
-    # Create a new run and log every provided parameter into it.
+
+def _run_entry_point(command, work_dir, active_run):
+    """
+    Runs an entry point command in a subprocess, waits on the subprocess to finish, & reports its
+    status to the tracking server.
+    """
+    run_id = active_run.run_info.run_uuid
+    eprint("=== Running command '%s' in run with ID '%s' === " % (command, run_id))
+    process = subprocess.Popen(["bash", "-c", command], close_fds=True, cwd=work_dir)
+    return LocalSubmittedRun(run_id, process)
+
+
+def _build_mlflow_run_cmd(uri, entry_point, storage_dir, use_conda, run_id, parameters):
+    mlflow_run_arr = ["mlflow", "run", uri, "-e", entry_point]
+    if storage_dir is not None:
+        mlflow_run_arr.extend(["--storage-dir", storage_dir])
+    if not use_conda:
+        mlflow_run_arr.append("--no-conda")
+    if run_id:
+        mlflow_run_arr.extend(["--run-id", run_id])
+    if parameters:
+        for key, value in parameters.items():
+            mlflow_run_arr.extend(["-P", "%s=%s" % (key, value)])
+    return mlflow_run_arr
+
+
+def _run_mlflow_run_cmd(mlflow_run_arr, env_map):
+    """
+    Invokes `mlflow run` in a subprocess using a special entry-point mode, which will in turn run
+    the entry point in a child process. Returns a handle to the subprocess.Popen launched to invoke
+    `mlflow run`.
+    """
+    final_env = os.environ.copy()
+    final_env.update(env_map)
+    # TODO: maybe write the output of asynchronous local runs e.g. as artifacts for
+    # debugging, we currently just drop it
+    # Launch `mlflow run` command as the leader of its own process group so that we can do a
+    # best-effort cleanup of all its descendant processes if needed
+    return subprocess.Popen(
+        mlflow_run_arr, env=final_env, universal_newlines=True,
+        stderr=open(os.devnull, "w"), stdout=open(os.devnull, "w"), preexec_fn=os.setsid)
```

Comment:
Was the previous behavior to interleave output, or drop the output? I kind of prefer the overly verbose strategy until we have a better story, otherwise it may be very hard to understand what goes wrong when things go wrong.

Reply:
Previously we dropped the output, I'm fine with having the output interleave